src/macros.rs

```diff
@@ -0,0 +1,323 @@
+/// Construct a `scale_value::Value`
+///
+///
+/// Supports unnamed and named composites and variants:
+/// ```
+/// use scale_value::value;
+///
+/// let val = value!({
+///     name: "localhost",
+///     address: V4(127, 0, 0, 1),
+///     payload: {
+///         bytes: (255, 3, 4, 9),
+///         method: ("Post", 3000),
+///     },
+/// });
+///
+/// ```
+/// Values can be nested in each other:
+/// ```
+/// use scale_value::value;
+///
+/// let data_value = value!((1, v1(1, 2), 3));
+/// let val = value!(POST { data: data_value });
+/// ```
+/// Trailing commas are optional.
+#[macro_export(local_inner_macros)]
+macro_rules! value {
+    ($($tt:tt)*) => {
+        value_internal!($($tt)*)
+    };
+}
+
+/// All patterns can be grouped into 4 main patterns:
+///
+/// ### `value_internal!(@unnamed [ELEMENTS] (REST))`
+///
+/// checks `REST` for certain patterns and if a suitable element pattern is found,
+/// the element is added to the comma seperated `ELEMENTS` list. When `REST` is empty,
+/// we collect all `ELEMENTS` into a Vec of values.
+///
+/// ### `value_internal!(@named [KEYVALUEPAIRS] (PARTIALKEY) (REST))`
+///
+/// goes over the REST tokens, to build up the PARTIALKEY until it is a proper KEY.
+/// This happens as soon as a colon `:` token is encountered, then we switch to the next pattern:
```

Comment:
Aah so this is the one that supports eg `value!({ hello there: 123 }` and aggregates "hello there" into a key?

I'd lean towards actually not supporting that (because it's a bit of a weird syntax, and shouldn't be needed anyways. (if it is, somebody can manually construct a value anyway).

If possible I'd support `value!({ "Hello there": 123 }` ie string keys, but I gather that'd be hard so I don't think it's necessary!

Reply:
The idea here was to build up any key value pair in two steps, each using a token muncher. First munch in all tokens into an initially empty pattern `()` and as soon as a colon is encountered, we put all of the tokens `(...)` patterns into a `[...]` to indicate the key is done and we are now inside of the value tokens. It is basically just a boolean switch that gives us context about where we are in the parsing. 
I did not have multi-part keys in mind when writing this and maybe I could go with an approach that just checks for a simple expression as a key, such that this "switch" is no longer needed.
About the string keys I need to check if its possible.